crates/external_services/src/grpc_client/dynamic_routing.rs

```diff
@@ -0,0 +1,213 @@
+    GrpcServerResponseFailure,
+}
+
+// Struct consists of all the services provided by the client
+#[derive(Debug, Clone)]
+pub struct RoutingStrategy {
+    pub success_rate_client: Option<SuccessRateCalculatorClient<Channel>>,
+}
+
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, Default)]
+pub struct DynamicRoutingClientConfig {
+    pub host: String,
+    pub port: u16,
+    pub enabled: bool,
+}
+
+// establish connection with the server
+impl DynamicRoutingClientConfig {
+    pub async fn get_dynamic_routing_connection(
+        self,
+    ) -> Result<RoutingStrategy, Box<dyn std::error::Error>> {
+        let success_rate_client = if self.enabled {
+            let uri = format!("http://{}:{}", self.host, self.port);
+            let channel = tonic::transport::Endpoint::new(uri)?.connect().await?;
+            Some(SuccessRateCalculatorClient::new(channel))
+        } else {
+            None
+        };
+        Ok(RoutingStrategy {
+            success_rate_client,
+        })
+    }
+}
+
+// The trait Success Based Dynamic Routing would have the functions required to support the calculation and updation window
+#[async_trait::async_trait]
+pub trait SuccessBasedDynamicRouting: dyn_clone::DynClone + Send + Sync {
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DRResult<CalSuccessRateResponse>;
+
+    async fn update_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        response: Vec<RoutableConnectorChoiceWithStatus>,
+    ) -> DRResult<UpdateSuccessRateWindowResponse>;
+}
+
+#[async_trait::async_trait]
+impl SuccessBasedDynamicRouting for SuccessRateCalculatorClient<Channel> {
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DRResult<CalSuccessRateResponse> {
+        let params = dynamic_routing_config
+            .params
+            .map(|vec| {
+                if !vec.is_empty() {
+                    vec.into_iter()
+                        .map(|param| param.to_string())
+                        .collect::<Vec<_>>()
+                        .join(":")
+                } else {
+                    String::default()
+                }
+            })
+            .get_required_value("Vector of params")
+            .change_context(DRError::ClientBuildingFailed)?;
+
+        let labels = label_input
+            .into_iter()
+            .map(|conn_choice| conn_choice.to_string())
+            .collect::<Vec<_>>();
+
+        let config = dynamic_routing_config.config.map(ForeignFrom::foreign_from);
+
+        let request = tonic::Request::new(CalSuccessRateRequest {
+            id: id.get_string_repr().to_owned(),
```

Comment:
Should we also have `tenant_id` prefix to id here so that all the dynamic routing id's related to one tenant is logically grouped